File: competition/ImageFiltering.py
import cv2
import numpy as np
import printer

'''
A class that prepares and filters an image.
'''
class ImageFiltering:
	def filterImage(self, img, constants):
		# Check if the image is None first
		if img is None:
			printer.printIfNeeded("Error: Image is empty!", constants)
		else:
			# Blur the image
			img = cv2.GaussianBlur(img, (5, 5), 0)
			
			# HSV filter the image
			img = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
			lower_range = np.array([constants.getValue("lowerh"), constants.getValue("lowers"), constants.getValue("lowerv")])
			higher_range = np.array([constants.getValue("higherh"), constants.getValue("highers"), constants.getValue("higherv")])
			HSVmask = cv2.inRange(img, lower_range, higher_range)
			img = cv2.bitwise_and(img, img, mask=HSVmask)
			img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

		return img

File: c++/ImageFiltering.cpp
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

class ImageFiltering {
	public:
	void filterImage(cv::Mat &img, cv::Mat &newimg) {
		clock_t t;
		int lowerh = 50;
		int lowers = 200;
		int lowerv = 30;
		int higherh = 66;
		int highers = 255;
		int higherv = 255;

		cv::GaussianBlur(img, img, cv::Size(5, 5), 0, 0);
	
		cv::cvtColor(img, img, CV_BGR2HSV);
	
		cv::Mat HSVmask;
	
		t = clock();
		cv::inRange(img, cv::Scalar(lowerh, lowers, lowerv), cv::Scalar(higherh, highers, higherv), HSVmask);
		t = clock() - t;
		std::cout << "ImageFiltering (inRange) - Time elapsed (s): " << ((float)t)/CLOCKS_PER_SEC << "\n";
		
		cv::bitwise_and(img, img, newimg, HSVmask);
		cv::imshow("Test", newimg);

		cv::cvtColor(newimg, newimg, CV_BGR2GRAY);	
	}
};
